### PR TITLE
Lexer and Parser for a very basic subset of the language

### PR DESCRIPTION
**Pull request by dalmemail**
Ready to merge, as we need to start with other phases of the compiler soon. I'll fix src/main.cpp and integrate a nicer test suite into cmake after that.

### CODE REVIEW

File: compiler/syntax/SLCParser.g4 [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    | VAR IDENTIFIER COLON type
+    ;
+
+declaration_list
+    : declaration
+    | declaration COMMA declaration_list
+    ;
+
+declaration_statement
+    : VAR declaration SEMICOLON
```
Comment (palmenros) [inline compiler/syntax/SLCParser.g4]: As declaration also starts with var, couldn't this grammar version allow `var var i : i32`? You should also allow constant variable declarations such as `const i : i32;`. Finally, we should also support a declaration with assignment where we specify the type such as `var i : i32 = 0;

Maybe we should split into different grammar rules function declarations and variable / constant declarations.

File: compiler/syntax/SLCParser.g4 [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+statement_list
+    : (statement)+
+    ;
+
+expression
+    : OPEN_PARENS expression CLOSE_PARENS
+    | expression2
+    ;
+
+expression2
```
Comment (palmenros) [inline compiler/syntax/SLCParser.g4]: To have a cleaner parse tree maybe we could use ANTLR 4 left recursion and operator precedence rules?

See The Definitive ANTLR 4 Reference, 5.4 Dealing with Precedence, Left Recursion, and Associativity and decide whether we should use that syntax.

File: compiler/syntax/SLCParser.g4 [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+declaration_list
+    : declaration
+    | declaration COMMA declaration_list
+    ;
+
+declaration_statement
+    : VAR declaration SEMICOLON
+    | VAR IDENTIFIER ASSIGN_OP expression SEMICOLON
+    ;
+
```
Comment (palmenros) [inline compiler/syntax/SLCParser.g4:49]: Are we allowing expressions as statements? This is a language design choice that we should consider carefully.

For example, is `var j=0; j;` a valid program in our language? Should we allow somehow reading global variables (such as hardware registers) that may result in behaviour change?

Maybe it's better to force to assign to `_` if we want to use an expression as statement, but we should consider it. 

Edit: thinking it twice `j++;` is an expression we absolutely want to support as an statement.

File: compiler/src/main.cpp [hunk 4]
```diff
@@ -29,23 +31,33 @@
...
+{
+    // temporary assume 4kb it's enough
+    char buffer[4096];
+    if (argc < 2)
+	return -1;
 
-//    ANTLRInputStream input("a = b + \"c\";(((x * d))) * e + f; a + (x * (y ? 0 : 1) + z);");
-    ANTLRInputStream input("a = pedro; c = h + d * b;");
-    TLexer lexer(&input);
+    int fd = open(argv[1], O_RDONLY);
```
Comment (palmenros) [inline compiler/src/main.cpp]: Please do not use platform-dependent code such as `open`. Instead use `fopen` or even better, you can use `ifstream` to not have to predeclare a buffer size. See https://stackoverflow.com/questions/18816126/c-read-the-whole-file-in-buffer 

For now we can use the manual approach, later we will let LLVM handle the source files for caching, includes, etc.

**Review by palmenros — CHANGES_REQUESTED** [review]
Nice job! Please see the comments attached.

File: compiler/syntax/SLCParser.g4 [hunk 5]
```diff
@@ -0,0 +1,141 @@
+// the name is temporary
+parser grammar SLCParser;
+options { tokenVocab=SLCLexer; }
+
+// https://github.com/antlr/antlr4/issues/606
+start
+    : program
```
Comment (palmenros) [inline compiler/syntax/SLCParser.g4]: Change `EOF` to `start` instead of `program` rule

File: compiler/syntax/SLCParser.g4 (hunk 1 above)
Comment (dalmemail) [reply compiler/syntax/SLCParser.g4]: I leave the const declarations out for this first version of the language

**Review by palmenros — APPROVED** [review]
Everything seems OK, great job!

**Pull request merged**